### PR TITLE
Feature/force exception refresh

### PR DESCRIPTION
**Pull request by p0psicles**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

image

@OmgImAlexis @ratoaq2 can you guys please check the route ivé added, to get scene exception last refresh, and force an update? I expect that we need to make changes to that.

Currently i'm using: ```api/v2/manage/searches/sceneexception/retrieve```
For the get and post (starting the syncing from source).

### CODE REVIEW

File: medusa/server/api/v2/scene_exception.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        last_updates = {}
+        for scene_exception_source in ['custom_exceptions', 'xem', 'anidb']:
+            last_updates[scene_exception_source] = get_last_refresh(scene_exception_source)[0]['last_refreshed']
+        return last_updates
+
+    def get(self):
+        """Return statistical information on the scene_exceptions table."""
+        return self.api_finish(data={'last_update': self.get_last_updates()})
+
+    def post(self):
```
Comment (OmgImAlexis) [inline medusa/server/api/v2/scene_exception.py]: `POST` is for sending data to be saved. @ratoaq2 do we have anything for RPC calls?

File: static/js/manage/manage-searches.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+MEDUSA.manage.manageSearches = function() {
+    // Get total number current scene exceptions per source. Will request medusa, xem and anidb name exceptions.
+    var url = 'manage/searches/sceneexception/retrieve';
+    var updateExceptionTable = function(exceptions) {
+        var status = $('#sceneExceptionStatus');
+
+        var table = $('<ul></ul>')
```
Comment (OmgImAlexis) [inline static/js/manage/manage-searches.js]: `.append`, `.on`, etc. should be on the same line as the ending of the last.

File: static/js/manage/manage-searches.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                new Date(exceptions.data.last_update.xem * 1000).toLocaleDateString())
+            .append('<li>Last updated anidb exceptions ' +
+                new Date(exceptions.data.last_update.anidb * 1000).toLocaleDateString());
+        status.append(table);
+
+        $('.forceSceneExceptionRefresh').removeClass('disabled');
+    };
+
+    api.get(url)
+        .then(function(response) {
```
Comment (OmgImAlexis) [inline static/js/manage/manage-searches.js]: Same here.

File: medusa/server/api/v2/scene_exception.py (hunk 1 above)
Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: Yes I was also wondering. Post looked to be the best option at the time.

Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: I do hope this doest need to become a really long discussion, and that we can be practical about it. Yes we've choose for rest. But these  rpc calls don't perform a clear action on a single resource. We could add sceneexceptions as a resource. But then there should be some room to run these kind of rpc.

**Comment by duramato** [comment]
What are the "costum exceptions"? Ours? I'd name them like Global exceptions or "Community Exceptions"
I personally prefer my 2nd suggestion.

**Comment by fernandog** [comment]
"Community Exceptions" - never heard of it =)

**Comment by duramato** [comment]
- Community made exceptions
- Community contributed exceptions
- Blabla

All better and more descriptive than "Costum exceptions"

File: static/js/manage/manage-searches.js (hunk 2 above)
Comment (p0psicles) [inline static/js/manage/manage-searches.js]: I tried to improve it. But some lines need to be broken, to prevent them from getting too long.

File: static/js/manage/manage-searches.js (hunk 3 above)
Comment (p0psicles) [inline static/js/manage/manage-searches.js]: This one I could continue

**Comment by p0psicles** [comment]
```Medusa's Scene Exceptions```

File: static/js/manage/manage-searches.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            "Last updated medusa's exceptions</a> " +
+            new Date(exceptions.data.last_update.custom_exceptions * 1000)
+                .toLocaleDateString()).append('<li>' +
+            '<a href="' + MEDUSA.config.anonRedirect +
+            'http://thexem.de">' +
+            'Last updated xem exceptions</a> ' +
+            new Date(exceptions.data.last_update.xem * 1000)
+                .toLocaleDateString()).append('<li>Last updated anidb exceptions ' +
+            new Date(exceptions.data.last_update.anidb * 1000)
+                .toLocaleDateString());
```
Comment (p0psicles) [inline static/js/manage/manage-searches.js]: @OmgImAlexis i'm not sure how to format this. I'd personally start every .append on a new line.

File: static/js/manage/manage-searches.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        status.append(table);
+
+        $('.forceSceneExceptionRefresh').removeClass('disabled');
+    };
+
+    api.get(url).then(function(response) {
+        updateExceptionTable(response);
+    }).catch(function(err) {
+        console.log('Trying to get scene exceptions failed with error: ' + err);
```
Comment (OmgImAlexis) [inline static/js/manage/manage-searches.js]: This should be using `log.error()` instead of `console.log()`.

File: static/js/manage/manage-searches.js (hunk 4 above)
Comment (OmgImAlexis) [inline static/js/manage/manage-searches.js]: Appends are annoying like this. Do it whichever way looks best to you and is readable.

Personally I'd set `new Date(exceptions.data.last_update.xem * 1000).toLocaleDateString()` and others to a var so it's easier to see what the append is doing.

**Comment by ratoaq2** [comment]
IMO, we should have a generic resource called operations, initially as follow:
```
{ type: REFRESH, created: 1970-01-01, user: 'super_user'}
```

The resource `sceneexception` should have operations as a sub-resource:
`sceneexception/operation`

Initially, we could support only creation:
```
PUT api/v2/sceneexception/operation
{ type: REFRESH}

HTTP RESPONSE
HTTP 201 CREATED
```

Later, if we decide so, we could persist these operations and search and list them using `GET /api/v2/sceneexception/operation`


This very same sub-resource could be used in other entities/resources as `application/operation` (to restart, shutdown, upgrade, etc), or a TVShow to update with indexer data:
```
updating the full series
PUT series/tvdb1234/operation
{ type: INDEXER_REFRESH}
```

```
updating only season 9
PUT series/tvdb1234/S09/operation
{ type: INDEXER_REFRESH}
```

In summary:

For this endpoint I'll go with:
```
PUT api/v2/sceneexception/operation
{ type: REFRESH}

HTTP RESPONSE
HTTP 201 CREATED
```

**Comment by p0psicles** [comment]
@ratoaq2 i'd like to get the last_update for every scene exception refresh. Shoul I requests that like 
```GET api/v2/sceneexception/lastupdate``` ?
As it's a different table the the scene_exceptions them self.

**Comment by ratoaq2** [comment]
We never did like that. For a specific scene exception you could:
`GET api/v2/sceneexception/{scene_exception_id_here}/lastupdate`

But for all scene exceptions, we usually use detailed=True/False
`GET api/v2/sceneexception/?detailed=False`
where it returns a detailed or limited information for each item.

I would recommend you to use `detailed=False`, so it would return all sceneexceptions but with fewer information, and of course the lastupdate field should be included

**Comment by p0psicles** [comment]
I've used the routes:
```GET api/v2/sceneexception/?detailed=False``` To get the last updates for all.
And
```PUT api/v2/sceneexception/operation
{ type: REFRESH}

HTTP RESPONSE
HTTP 201 CREATED
```
To force the operation.

**Comment by p0psicles** [comment]
Used some code from #1787 for the sceneexception route.

**Comment by fernandog** [comment]
tested and no issues. 
wondering if we can add time in the UI for "last updated"

**Comment by p0psicles** [comment]
Yeah, we could. but you have to ask yourself, would that matter? It's not that these scene exceptions are updated every hour.

**Comment by fernandog** [comment]
@p0psicles yes, it was just a confirm that current time = last update time. No biggy

File: medusa/server/api/v2/scene_exception.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+class SceneExceptionOperationHandler(BaseRequestHandler):
+    """Scene Exception request handler."""
+
+    def set_default_headers(self):
+        """Set default CORS headers."""
+        super(SceneExceptionOperationHandler, self).set_default_headers()
+        self.set_header('Access-Control-Allow-Methods', 'PUT')
+
+    def put(self):
```
Comment (OmgImAlexis) [inline medusa/server/api/v2/scene_exception.py]: Should this not be POST to be similar to how @p0psicles @ratoaq2 and I discussed?

POST is for making new resources where as PUT is for replacing a resource.

File: medusa/server/api/v2/scene_exception.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+        exceptions = [{'exception_id': row[0],
+                       'indexer': row[1],
+                       'indexer_id': row[2],
+                       'show_name': row[3],
+                       'season': row[4],
+                       'custom': row[5]}
+                      for row in exceptions]
+
+        self.api_finish(data={"exceptions": exceptions, "last_updates": get_last_updates()})
```
Comment (OmgImAlexis) [inline medusa/server/api/v2/scene_exception.py]: All JSON returned should be in camelCase.

File: medusa/server/api/v2/scene_exception.py (hunk 6 above)
Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: I changed to put, on rats examples.

Comment (OmgImAlexis) [inline medusa/server/api/v2/scene_exception.py]: @ratoaq2 any reason for this?

File: medusa/server/api/v2/scene_exception.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        :param season:
+        """
+        exception_id = self._parse(kwargs.pop('row_id'))
+        indexer = self.get_query_argument('indexer', None)
+        indexer_id = self.get_query_argument('indexer_id', None)
+        season = self.get_query_argument('season', None)
+        detailed = self.get_query_argument('detailed', 'true').lower()
+
+        if detailed == 'false':
+            return self.api_finish(data={"last_updates": get_last_updates()})
```
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: GET sceneexception should always return sceneexception. Detailed will return with all info. Non detailed should return with just `header` info

File: medusa/server/api/v2/scene_exception.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+    def get(self, *args, **kwargs):
+        """Query scene_exception information.
+
+        :param show_indexer_id: The indexer id. Like 1 for tmdb and 3 for tvmaze.
+        :param show_id:
+        :type show_indexer_id: str
+        :param season:
+        """
+        exception_id = self._parse(kwargs.pop('row_id'))
```
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: When your requesting a resource by its id, you should just return it. Do not return a singleton list, but just the resource itself

File: medusa/server/api/v2/scene_exception.py (hunk 7 above)
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: And it should be a simple list of sceneexception. Here it's a combination of sceneexception and timestamps

Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: Do you need to have timestamp per indexer? otherwise you could just use last modified http header

File: medusa/server/api/v2/scene_exception.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+                            b', season = ?'
+                            b', custom = 1'
+                            b' WHERE exception_id = ?',
+                            [data.get('indexer'),
+                             data.get('indexer_id'),
+                             data.get('show_name'),
+                             data.get('season'),
+                             exception_id])
+        if cache_db_con.connection.total_changes - last_changes == 1:
+            return self.api_finish(status=201)
```
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: Http 201 created. In this case it was updated. either you return 200 with the exception itself in the http body, or just (preferably) 204 http no content

File: medusa/server/api/v2/scene_exception.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        last_changes = cache_db_con.connection.total_changes
+        cache_db_con.action(b'INSERT INTO scene_exceptions'
+                            b' (indexer, indexer_id, show_name, season, custom) '
+                            b' values (?,?,?,?,1)',
+                            [data.get('indexer'),
+                             data.get('indexer_id'),
+                             data.get('show_name'),
+                             data.get('season')])
+        if cache_db_con.connection.total_changes - last_changes > 0:
+            return self.api_finish(status=204)
```
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: You're creating a resource. You should return it back in the response since new fields could have been populated (auto generated ids, etc). 
Either you return 201 created with http header with the url location for the resource or http 200 with the resource itself in the response body

File: medusa/server/api/v2/scene_exception.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                            b', custom = 1'
+                            b' WHERE exception_id = ?',
+                            [data.get('indexer'),
+                             data.get('indexer_id'),
+                             data.get('show_name'),
+                             data.get('season'),
+                             exception_id])
+        if cache_db_con.connection.total_changes - last_changes == 1:
+            return self.api_finish(status=201)
+        return self.api_finish(status=400, error="Could not update.")
```
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: If resource is not found, 404 should be returned

File: medusa/server/api/v2/scene_exception.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        :type exception_id: str
+        """
+        exception_id = self._parse(kwargs.pop('row_id'))
+
+        cache_db_con = db.DBConnection('cache.db')
+        last_changes = cache_db_con.connection.total_changes
+        cache_db_con.action(b'DELETE FROM scene_exceptions WHERE exception_id = ?', [exception_id])
+        if cache_db_con.connection.total_changes - last_changes > 0:
+            return self.api_finish(status=204)
+        return self.api_finish(status=400, error="Failed to delete.")
```
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: If resource is not found, 404 should be returned

File: medusa/server/api/v2/scene_exception.py (hunk 9 above)
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: And if the resource is not found, 404 should be returned

File: medusa/server/core.py [hunk 14]
```diff
@@ -23,22 +23,40 @@ def get_apiv2_handlers(base):
...
 
     return [
-        (r'{base}/show(?:/{show_id}(?:/{ep_id})?(?:/{query})?)?/?'.format(base=base, show_id=show_id, ep_id=ep_id, query=query), ShowHandler),
-        (r'{base}/config(?:/{query})?/?'.format(base=base, query=query), ConfigHandler),
-        (r'{base}/log(?:/{log_level})?/?'.format(base=base, log_level=log_level), LogHandler),
-        (r'{base}/authenticate(/?)'.format(base=base), AuthHandler),
-        (r'{base}/asset(?:/{asset_group})(?:/{query})?/?'.format(base=base, asset_group=asset_group, query=query_extended), AssetHandler),
-        (r'{base}(/?.*)'.format(base=base), NotFoundHandler)
+        # All operations endpoints should be defined first.
+        (r'{base}/sceneexception/operation'.format(base=base),
```
Comment (ratoaq2) [inline medusa/server/core.py]: I need to think about this and I'll come back to you. Operations on a single resource or in a resource list. This was it's on a resource list

File: medusa/server/api/v2/scene_exception.py (hunk 8 above)
Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: Ok, so when I need the last_updates. I pass with detailed. And I return both?

Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: what do you mean with: `Non detailed should return with just header info`

File: medusa/server/api/v2/scene_exception.py (hunk 11 above)
Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: Then i'd prefer to return the resource with a 200.

File: medusa/server/api/v2/scene_exception.py (hunk 7 above)
Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: I have the lastupdates stored per source. Also when one source is down for ex. The user can see this, because two have latest date, and one is lagging 3 days behind. So maybe that's a resource in itself? Like `sceneexceptionupdate/xem` `sceneexceptionupdate/custom` `sceneexceptionupdate/anidb`?
But where do you put the operation?

**Comment by fernandog** [comment]
Implements https://github.com/pymedusa/Medusa/issues/410

File: medusa/server/api/v2/scene_exception.py (hunk 7 above)
Comment (labrys) [inline medusa/server/api/v2/scene_exception.py]: JSON has no requirement that keys be camelCased.  Since our API keys directly reference a Python back-end they should remain snake_cased.

File: medusa/server/api/v2/scene_exception.py (hunk 6 above)
Comment (labrys) [inline medusa/server/api/v2/scene_exception.py]: Actually you can use PUT or POST to create resources.  PUT **MUST** be idempotent, POST is **NOT** idempotent.

File: medusa/server/api/v2/scene_exception.py (hunk 9 above)
Comment (labrys) [inline medusa/server/api/v2/scene_exception.py]: If it is always expected to be a single item, then yes it should return a single item and not a list containing an item.  If a list of items is expected, it should always return a list, even if it only contains a single item, (or no items).

File: medusa/server/api/v2/scene_exception.py (hunk 6 above)
Comment (OmgImAlexis) [inline medusa/server/api/v2/scene_exception.py]: @labrys PUT should not be used as the main way of adding new resources.

File: medusa/server/api/v2/scene_exception.py (hunk 7 above)
Comment (OmgImAlexis) [inline medusa/server/api/v2/scene_exception.py]: camelCase is what we have as our standard over all of the API. Please use it.

File: medusa/server/api/v2/scene_exception.py (hunk 6 above)
Comment (labrys) [inline medusa/server/api/v2/scene_exception.py]: http://dret.typepad.com/dretblog/2011/11/creating-resources-with-get-put.html
"creating resources the RESTful way can be done with either PUT or POST, with PUT expecting that the URI of the created resource is known in advance (by the client issuing the request)"  

http://restcookbook.com/HTTP%20Methods/put-vs-post/
"The HTTP methods POST and PUT aren't the HTTP equivalent of the CRUD's create and update. They both serve a different purpose. It's quite possible, valid and even preferred in some occasions, to use PUT to create resources, or use POST to update resources."

https://jcalcote.wordpress.com/2008/10/16/put-or-post-the-rest-of-the-story/
"Create can be implemented using an HTTP PUT, if (and only if) the payload of the request contains the full content of the exactly specified URL."

Comment (labrys) [inline medusa/server/api/v2/scene_exception.py]: https://tools.ietf.org/html/rfc2616#section-9.6
> The PUT method requests that the enclosed entity be stored under the
>    supplied Request-URI. If the Request-URI refers to an already
>    existing resource, the enclosed entity SHOULD be considered as a
>    modified version of the one residing on the origin server. If the
>    Request-URI does not point to an existing resource, and that URI is
>    capable of being defined as a new resource by the requesting user
>    agent, the origin server can create the resource with that URI. If a
>    new resource is created, the origin server MUST inform the user agent
>    via the 201 (Created) response. If an existing resource is modified,
>    either the 200 (OK) or 204 (No Content) response codes SHOULD be sent
>    to indicate successful completion of the request.

File: medusa/server/api/v2/scene_exception.py [hunk 15]
```diff
@@ -16,20 +16,67 @@ def get_last_updates():
...
+class SceneExceptionTypeHandler(BaseRequestHandler):
+    """Scene Exception type request handler."""
+
+    def set_default_headers(self):
+        """Set default CORS headers."""
+        super(SceneExceptionTypeHandler, self).set_default_headers()
+        self.set_header('Access-Control-Allow-Methods', 'GET')
+
+    def get(self, exception_type):
+        if exception_type is not None and exception_type not in ['medusa', 'xem', 'anidb']:
```
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: I would go with `custom` instead of using the application name (medusa)

Comment (labrys) [inline medusa/server/api/v2/scene_exception.py]: While we're talking about naming exceptions, instead of `SceneExceptions` what do you think about calling them `NamingExceptions`?  They don't only apply to scene, and they aren't exceptions to Scene, so it might be a little more appropriate.

File: medusa/server/api/v2/scene_exception.py [hunk 16]
```diff
@@ -16,20 +16,67 @@ def get_last_updates():
...
+        super(SceneExceptionAllTypeOperationHandler, self).set_default_headers()
+        self.set_header('Access-Control-Allow-Methods', 'POST')
 
     def post(self):
-        """Start fetch retrieving scene name exceptions."""
+        """Start scene exceptions type operation. To refresh all scene exception types."""
+        json_body = json_decode(self.request.body)
+
+        if json_body.get('type', '') == 'REFRESH':
+            retrieve_exceptions(force=True, exception_type=None)
```
Comment (OmgImAlexis) [inline medusa/server/api/v2/scene_exception.py]: This starts an update, right? Should this not return something like in progress?

@ratoaq2 not sure what return we'd use for something that isn't "complete".

File: medusa/server/api/v2/scene_exception.py (hunk 15 above)
Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: I'd like to reserve a type for manual added exceptions. Thought I could use custom for that. Then use Medusa for ours.

Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: I think changing to naming exception is fine. But then should be changed in a big bang. Db, api, code, comments, wiki. So don't think this Pr is really a good location to discuss that.

File: medusa/server/api/v2/scene_exception.py (hunk 16 above)
Comment (ratoaq2) [inline medusa/server/api/v2/scene_exception.py]: Is it asynchronous? If so, 202 ACCEPTED should be used. Everything else looks good.

Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: It's a synchronous backend process.

File: static/js/manage/manage-searches.js [hunk 17]
```diff
@@ -2,11 +2,19 @@ MEDUSA.manage.manageSearches = function() {
     // Get total number current scene exceptions per source. Will request medusa, xem and anidb name exceptions.
     var updateExceptionTable = function(exceptions) {
         var status = $('#sceneExceptionStatus');
-        var medusaException = _.find(exceptions.data, function(obj) { return obj.id === 'medusa' })
+        var medusaException = _.find(exceptions.data, function(obj) {
```
Comment (OmgImAlexis) [inline static/js/manage/manage-searches.js]: Why not use the built in js filter function?

```
exceptions.data.filter(function(obj) {
    return obj.id === 'medusa';
});
```

**Comment by OmgImAlexis** [comment]
Since this was swallowed as an outdated commit.


Why not use the built in js filter function instead of the `_.find()`?
```
exceptions.data.filter(function(obj) {
    return obj.id === 'medusa';
});
```

**Comment by p0psicles** [comment]
@labrys can you update your review? Lot's has changed and discussed since then.

File: medusa/server/api/v2/scene_exception.py (hunk 16 above)
Comment (p0psicles) [inline medusa/server/api/v2/scene_exception.py]: Put in a spinner to get some visual feedback.

**Comment by p0psicles** [comment]
If all good I'll merge this.
